File: numconverter.java
import java.util.*;
public class numconverter {
    public static void main(String[] args) {
        while (true){
            //The Scanner syntax to take input
            Scanner sc = new Scanner(System.in);
            // Declaring all the variables
            String num, hexalpha = "abcdef0123456789";
            int base, lenofnum, sum_of_digits = 0;
            boolean isint = true, isHex = false;
            /*Interface to accept the number and its base 
            and store it's value*/
            System.out.print("Enter a number: ");
            num= sc.next();
            System.out.print("Enter base of number(2,8,10,16): ");
            base=sc.nextInt();
            //Finding the number of digits
            lenofnum = num.length();
            //Checking if base is valid
            if (base == 2 ||base == 8 ||base == 10 ||base == 16) {
                //Checking if the number is an integer
                for (int i = 0; i < num.length(); i++) {
                    try {
                        Integer.parseInt(num.charAt(i)+"");
                    } catch (Exception NumberFormatException) {
                        isint = false;
                    }
                    //Finding the sum of the digits of the number
                    if (isint==true) {
                        sum_of_digits+=Integer.parseInt(num.charAt(i)+"");
                    }
                    // Checking if number is a valid Hexadecimal number
                    if (hexalpha.indexOf((num.charAt(i)+"").toLowerCase())!=-1) {
                        isHex = true;
                    }else {
                        isHex = false;
                        break;
                    }
                }
            }
            //Checking if the number if a Binary number and converting it to
            if (base==2 && (sum_of_digits/lenofnum)<=1 && isint==true) {
                // An Octal Number
                System.out.println("Octal Digit is: ("+decimal_to_other(other_to_decimal(num,2),8)+") [base: 8]");
                // A Decimal number
                System.out.println("Decimal Digit is: ("+other_to_decimal(num,2)+") [base: 10]");
                // A Hexadecimal Number
                System.out.println("Hexadecimal Digit is: ("+decimal_to_other(other_to_decimal(num,2),16)+") [base: 16]");
            //Checking if the number if an Octal number and converting it to
            }else if (base==8 && (sum_of_digits/lenofnum)<=7 && isint==true) {
                // A Binary Number
                System.out.println("Binary Digit is: ("+decimal_to_other(other_to_decimal(num,8),2)+") [base: 2]");
                // A Decimal Number
                System.out.println("Decimal Digit is: ("+other_to_decimal(num,8)+") [base: 10]");
                // A Decimal Number
                System.out.println("Hexadecimal Digit is: ("+decimal_to_other(other_to_decimal(num,8),16)+") [base: 16]");
            //Checking if the number if a Decimal number and converting it to
            }else if (base==10 && (sum_of_digits/lenofnum)<=9 && isint==true) {
                // A Binary Number
                System.out.println("Binary Digit is: ("+decimal_to_other(num,2)+") [base: 2]");
                // An Octal Number
                System.out.println("Octal Digit is: ("+decimal_to_other(num,8)+") [base: 8]");
                // A Hexadecimal Number
                System.out.println("Hexadecimal Digit is: ("+decimal_to_other(num,16)+") [base: 16]");
            //Checking if the number if a Hexadecimal number and converting it to
            }else if (base==16 && isHex==true) {
                // A Binary Number
                System.out.println("Binary Digit is: ("+decimal_to_other(other_to_decimal(num,16),2)+") [base: 2]");
                // An Octal Number
                System.out.println("Octal Digit is: ("+decimal_to_other(other_to_decimal(num,16),8)+") [base: 8]");
                // A Decimal Number
                System.out.println("Decimal Digit is: ("+other_to_decimal(num,16)+") [base: 10]");
            }else{
                System.out.println("Not a valid base or number or both.");
            }
        }
    }
    /* Function to convert a decimal number and convert it to any base like 
    Binary, Octal, Decimal and Hexadecimal
    */ 
    public static String decimal_to_other(String number, int base_to_convert) {
        //Declaring all the important variables
        String answer = "",hexalpha, actual_answer="";
        hexalpha = "ABCDEF";
        char ch;
        //The main loop that will do two things
        while (Integer.parseInt(number)!=0) {
            //1. Find the modulus of number
            String moduluser="";
            moduluser = Integer.toString(Integer.parseInt(number)%base_to_convert);
            /* For Hexadecimal number:
             * Replaces the modulus with alphabets accordingly
             */
            if (base_to_convert==16) {
                for (int i=0;i<hexalpha.length();i++) {
                    if ((i+10)==Integer.parseInt(moduluser)) {
                        moduluser = hexalpha.charAt(i)+"";
                        break;
                    }
                }
            }
            //2. Add the modulus to the answer that is in reverse
            answer+=moduluser;
            number = Integer.toString(Integer.parseInt(number)/base_to_convert);
        }
        // Re-reversing the number to give the result
        for (int i=0; i<answer.length(); i++)
        {
            ch= answer.charAt(i);
            actual_answer= ch+actual_answer;
        }
        // Returning the answer
        return actual_answer;
    }
    /* Function to convert a number from any base like
    Binary, Decimal, Octal and Hexadecimal
    and convert it to a decimal number
    */ 
    public static String other_to_decimal(String num, int converter) {
        //Declaring all the important variables
        int lenofnum,index_of_digit,answer;
        boolean isint=true;
        String hexalpha;
        hexalpha = "ABCDEF";
        index_of_digit=0;
        answer=0;
        //Finds the number of digits of the number
        lenofnum = num.length()-1;
        /* The main loop for this function that does
        all the conversions
         */
        while (lenofnum!=-1) {
            // Checking if number is a Hexadecimal number
            if (converter==16) {
                // Dividing the number into numbers and alphabets
                try {
                    Integer.parseInt(num.charAt(index_of_digit)+"");
                } catch (Exception NumberFormatException) {
                    isint=false;
                }
                //Processing the alphabet part
                if (isint==false) {
                    if (hexalpha.indexOf((num.charAt(index_of_digit)+"").toUpperCase())!=-1) {
                        answer+=((Integer.parseInt(hexalpha.indexOf((num.charAt(index_of_digit)+"").toUpperCase())+"")+10)*((int)(Math.pow(converter, lenofnum))));
                    } else {//Or displaying that the Hexadecimal number is invalid
                        System.out.println("Not a valid Hexadecimal number. ");
                        break;
                    }
                // Processing the Integer part
                } else {
                    answer+=(Integer.parseInt(num.charAt(index_of_digit)+"")*(int)(Math.pow(converter,lenofnum)));
                }
            // Processing the numbers of other bases
            }else {
                answer+=(Integer.parseInt(num.charAt(index_of_digit)+"")*(int)(Math.pow(converter,lenofnum)));
            }
            // Incrementing and decrementing numbers
            index_of_digit++;
            lenofnum--;
        }
        // Returning the final answer
        return Integer.toString(answer);
    }
}

File: numconverter.py
import math
def decimal_to_other(number, base_to_convert):
    z = ''
    ya = {10: 'A', 11: 'B', 12:'C',13:'D',14:'E',15:'F'}
    while number/base_to_convert!=0:
        x = str(number%base_to_convert)
        if base_to_convert==16:
            if ya.get(int(x))!=None:
                x = ya.get(int(x)) 
        z = z+(x)
        number=int(number/base_to_convert)
    return z[::-1]
def other_to_decimal(num, converter):
    lenofnum = len(str(num))-1
    za = 0
    z=0
    ya = {'a': 10, 'A':10,'B':11,'b':11,'C':12,'c':12,'D':13,'d':13,'E':14,'e':14,'F':15,'f':15}
    while lenofnum!=-1:
        if converter==16:
            isint=True
            try:
                int(str(num)[za])
            except ValueError:
                isint=False
            if isint==False:
                if ya.get(str(num)[za]) !=None:
                    z+=(int(str(ya.get(str(num)[za])))*int(math.pow(converter, lenofnum)))
                else:
                    print("Not a valid Hexadecimal number.")
                    exit()
            else:
                z+=(int(str(num)[za])*int(math.pow(converter, lenofnum)))
        else:
            z+=(int(str(num)[za])*int(math.pow(converter, lenofnum)))
        za+=1
        lenofnum-=1
    return z
    
def main():
    zrz = 0
    hexalpha = "abcdef0123456789"
    num = input("Enter a number: ")
    base = int(input("Enter base of number(2,8,10,16): "))
    x=len(num)
    z=0
    isf = True
    isHex = False
    isint = True
    if base==2 or base==8 or base==10 or base==16:
        for i in range(len(num)):
            try:
                int(num[i])
            except ValueError:
                isint = False
            if isint==True:
                z+=int(num[i])
            try:
                hexalpha.index(num[i])
            except ValueError:
                isHex = False
                isf = False
            if num[i] == "0":
                zrz+=1
            if isf==True:
                isHex = True
            else:
                isHex = False
                break
    x = x-zrz
    if base==2 and (z/x)<=1 and isint==True:
        print("Octal Digit is: ("+str(decimal_to_other(other_to_decimal(num, 2), 8))+")8")
        print("Decimal digit is: ("+str(other_to_decimal(num, 2))+")10")
        print("Hexadecimal digit is: ("+str(decimal_to_other(other_to_decimal(num, 2), 16))+")16")
    elif base==8 and int(z/len(num))<=7 and isint==True:
        print("Binary Digit is: ("+str(decimal_to_other(other_to_decimal(num,8),2))+")2")
        print("Decimal Digit is: ("+str(other_to_decimal(num, 8))+")10")
        print("Hexadecimal Digit is: ("+str(decimal_to_other(other_to_decimal(num, 8), 16))+")16")
    elif base==10 and z/x<=9 and isint==True:
        print("Binary digit is: (" + str(decimal_to_other(int(num), 2))+")2")
        print("Octal digit is: (" + str(decimal_to_other(int(num), 8))+")8")
        print("Hexadecimal digit is: (" + str(decimal_to_other(int(num), 16))+")16")
    elif base==16 and isHex==True:
        print("Binary digit is: ("+str(decimal_to_other(other_to_decimal(num, 16), 2))+")2")
        print("Octal digit is: ("+str(decimal_to_other(other_to_decimal(num, 16), 8))+")8")
        print("Decimal digit is: ("+str(other_to_decimal(num,16))+")10")
    else:
        print("Not a valid base or number or both.")
        if base!=2 and base!= 8 and base!=10 and base!=16:
            idk = "base"
        else:
            idk = "number"
        x = "Do you want to see results for your invalid " + idk + "?: "
        stir = input(x)
        if stir[0].lower()=="y":
            print("\nThen Here you go: \n")
            print("Binary digit: " + str(decimal_to_other(other_to_decimal(num, base), 2)))
            print("Octal digit: " + str(decimal_to_other(other_to_decimal(num, base), 8)))
            print("Decimal digit: " + str(other_to_decimal(num, base)))
            print("Hexadecimal digit: " + str(decimal_to_other(other_to_decimal(num, base), 16)))
        else: 
            print("Ok, bye :(")
main()
